cmd/downsampler.go

```diff
@@ -0,0 +1,162 @@
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/digitalocean/vulcan/cassandra"
+	"github.com/digitalocean/vulcan/downsampler"
+	"github.com/digitalocean/vulcan/kafka"
+	"github.com/gocql/gocql"
+	hostpool "github.com/hailocab/go-hostpool"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+// Downsampler handles parsing the command line options, initializes, and starts the
+// downnsampler service accordingling.  It is the entry point for the downnsampler
+// service.
+func Downsampler() *cobra.Command {
+	d := &cobra.Command{
+		Use:   "downsampler",
+		Short: "consumes timeseries sent by the forwarder and stores them at the resolution configured to the timeseries storage",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var (
+				reg     = prometheus.NewRegistry()
+				signals = make(chan os.Signal, 1)
+			)
+
+			signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
+
+			// get kafka source
+			s, err := kafka.NewSource(&kafka.SourceConfig{
+				Addrs:    strings.Split(viper.GetString(flagKafkaAddrs), ","),
+				ClientID: viper.GetString(flagKafkaClientID),
+				GroupID:  viper.GetString(flagKafkaGroupID),
+				Topics:   []string{viper.GetString(flagKafkaTopic)},
+			})
+			if err != nil {
+				return err
+			}
+			//reg.MustRegister(s)
+			log.WithFields(log.Fields{
+				"kafka_client_id": viper.GetString(flagKafkaClientID),
+				"kafka_group_id":  viper.GetString(flagKafkaGroupID),
+				"kafka_topic":     viper.GetString(flagKafkaTopic),
+				"kafka_addresses": viper.GetString(flagKafkaAddrs),
+			}).Info("registered as kafka consumer")
+
+			// Setup Cassandra writer
+			cluster := gocql.NewCluster(strings.Split(viper.GetString(flagCassandraAddrs), ",")...)
+			cluster.Keyspace = viper.GetString(flagCassandraKeyspace)
+			cluster.Timeout = viper.GetDuration(flagCassandraTimeout)
+			cluster.NumConns = viper.GetInt(flagCassandraNumConns)
+			cluster.Consistency = gocql.LocalOne
+			cluster.ProtoVersion = 4
+			// Fallback simple host pool distributes queries and prevents sending queries to unresponsive hosts.
+			fallbackHostPolicy := gocql.HostPoolHostPolicy(hostpool.New(nil))
+			// Token-aware policy performs queries against a host responsible for the partition.
+			// TODO in gocql make token-aware able to write to any host for a partition when the
+			// replication factor is > 1.
+			// https://github.com/gocql/gocql/blob/4f49cd01c8939ce7624952fe286c3d08c4be7fa1/policies.go#L331
+			cluster.PoolConfig.HostSelectionPolicy = gocql.TokenAwareHostPolicy(fallbackHostPolicy)
+			sess, err := cluster.CreateSession()
+			if err != nil {
+				return err
+			}
+
+			tablename := fmt.Sprintf("downsampler_%v", viper.GetDuration(flagDownsamplerResolution))
```

Comment:
I don't think it's clear to the operator that `downsampler_${resolution}` will be the table name. Something we could iterate on later.